doc/design/cluster_train/pserver_client.md

```diff
@@ -0,0 +1,126 @@
+```
+The selected trainer's call to `paddle_begin_init_params` will return with 1, and the other trainers' call to `paddle_begin_init_params` will block until initialization is done, and return 0. As illustrated below:
+
+<img src="./src/pserver_init.png">
+
+## C Interface
+
+```c
+#define PADDLE_ELEMENT_TYPE_INT32   0
+#define PADDLE_ELEMENT_TYPE_UINT32  1
+#define PADDLE_ELEMENT_TYPE_INT64   2
+#define PADDLE_ELEMENT_TYPE_UINT64  3
+#define PADDLE_ELEMENT_TYPE_FLOAT32 4
+#define PADDLE_ELEMENT_TYPE_FLOAT64 5
+
+typedef struct {
+  char* name;
+  int   element_type;
+  void* content;
+  int   content_len;
+} paddle_parameter, paddle_gradient;
+
+typedef struct paddle_pserver_client paddle_pserver_client;
+
+paddle_pserver_client* paddle_new_pserver_client();
+void paddle_pserver_client_release(paddle_pserver_client* client);
+
+/**
+ * @brief paddle_begin_init_params begins to initialize parameters on
+ * parameter servers.
+ *
+ * paddle_begin_init_params will be called from multiple trainers,
+ * only one trainer will be selected to initialize the parameters on
+ * parameter servers. Other trainers will be blocked until the
+ * initialization is done, and they need to get the initialized
+ * parameters from parameter servers using @paddle_get_params.
+ *
+ * @param config_proto serialized parameter server configuration in
+ * Protocol Buffers format.
+ * @return 1 if the trainer is selected to initialize parameter
+ * servers, otherwise 0.
+ */
+int paddle_begin_init_params(paddle_pserver_client* client, const char* config_proto);
+
+/**
+ * @brief paddle_init_param initializes the parameter on parameter
+ * servers.
+ *
+ * @param param the parameter to initialize.
+ * @return 0 if successful, otherwise -1. On failure, the trainer
+ * needs to restart the entire initialization process (starting from
+ * @paddle_begin_init_param). Or simply exit the program and wait for
+ * the cluster management system to restart the trainer.
+ */
+int paddle_init_param(paddle_pserver_client* client, paddle_parameter params);
+
+/**
+ * @brief paddle_finish_init_params tells parameter servers client has
+ * sent all parameters to parameter servers as initialization.
+ *
+ * @return 0 if successful, otherwise -1. On failure, the trainer
+ * needs to restart the entire initialization process (starting from
+ * @paddle_begin_init_param). Or simply exit the program and wait for
+ * the cluster management system to restart the trainer.
+ */
+int paddle_finish_init_params(paddle_pserver_client* client);
+
+/**
+ * @brief paddle_send_grads sends gradients to parameter servers for
+ * updating parameters.
+ *
+ * @param grads the array of gradients to send.
+ * @param total the total number of gradient inside the gradient array.
+ * @param learning_rate the learning rate for the gradients.
+ * @return 0 if successful, otherwise -1.
+ */
+int paddle_send_grads(paddle_pserver_client* client, const paddle_gradient* grads, int total, double learning_rate);
```

Comment:
maybe use the same interface with "set_parameter" better?  otherwise, the set_parameters only used once.